### PR TITLE
[stdlib] add `__str__` and `__repr__` for `Optional`

### PR DESCRIPTION
**Pull request by bgreni**
Add `format_to` `__str__` and `__repr__`

### CODE REVIEW

File: stdlib/src/collections/optional.mojo [hunk 1]
```diff
@@ -154,6 +154,56 @@ struct Optional[T: CollectionElement](CollectionElement, Boolable):
...
+
+        Parameters:
+            U: The type of the elements in the list. Must implement the
+              traits `Representable` and `CollectionElement`.
+
+        Args:
+            writer: The formatter to write to.
+        """
+        if self:
+            write_to(writer, repr(self.value()))
```
Comment (gabrieldemarmiesse) [inline stdlib/src/collections/optional.mojo:203]: Here `repr(self.value())` can be a very big `String`. If we know that `self.value()` has a `format_to` method, would it be possible to use it to write directly to the `Formatter`?

You can imagine that `self.value()` is a List with 10 000 elements for example.

**Review by gabrieldemarmiesse — COMMENTED** [review]
Nice addition. I think we can be more efficient here by skipping a temporary allocation of a possibly big string.

File: stdlib/src/collections/optional.mojo (hunk 1 above)
Comment (bgreni) [reply stdlib/src/collections/optional.mojo:203]: I like the idea, but wouldn't that require use to generally shift away from `RepresentableCollectionElement` in these scenarios to something equivalent to `Formattable & CollectionElement`?

**Comment by ConnorGray** [comment]
!sync

**Comment by ConnorGray** [comment]
It looks like I can't sync this because of the merge conflict. @bgreni could you rebase and then I'll merge this? Thanks! 🙂

File: stdlib/src/collections/optional.mojo (hunk 1 above)
Comment (ConnorGray) [reply stdlib/src/collections/optional.mojo:203]: I agree with Gabriel that we could make this more efficient, though we can do that in a follow-up PR if you'd like 🙂

Comment (bgreni) [reply stdlib/src/collections/optional.mojo:203]: Might as well do it here. I'm on vacation at the moment but I can look into it on Monday!

**Comment by bgreni** [comment]
> It looks like I can't sync this because of the merge conflict. @bgreni could you rebase and then I'll merge this? Thanks! 🙂

Done now, apologies for the delay!

**Comment by ConnorGray** [comment]
> Done now, apologies for the delay!

No worries at all, thank you for fixing the conflict! I'll sync this in now.

**Comment by ConnorGray** [comment]
!sync

**Pull request closed without merging**